contracts/UniAgent.sol

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.17;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { TokenCollector } from "./abstracts/TokenCollector.sol";
+import { EIP712 } from "./abstracts/EIP712.sol";
+import { Ownable } from "./abstracts/Ownable.sol";
+import { IWETH } from "./interfaces/IWETH.sol";
+import { IUniAgent } from "./interfaces/IUniAgent.sol";
+import { Asset } from "./libraries/Asset.sol";
+import { Constant } from "./libraries/Constant.sol";
+import { SignatureValidator } from "./libraries/SignatureValidator.sol";
+
+contract UniAgent is Ownable, IUniAgent, TokenCollector, EIP712 {
+    using Asset for address;
+    using SafeERC20 for IERC20;
+
+    address private constant v2Router = 0x7a250d5630B4cF539739dF2C5dAcb4c659F2488D;
+    address private constant v3Router = 0xE592427A0AEce92De3Edee1F18E0157C05861564;
+    address payable private constant universalRouter = payable(0xEf1c6E67703c7BD7107eed8303Fbe6EC2554BF6B);
+
+    IWETH public immutable weth;
+    address payable public feeCollector;
+
+    mapping(bytes32 => bool) private filledOrder;
+
+    /// @notice Emitted when fee collector address is updated
+    /// @param newFeeCollector The address of the new fee collector
+    event SetFeeCollector(address newFeeCollector);
+
+    constructor(
+        address _owner,
+        address _uniswapPermit2,
+        address _allowanceTarget,
+        IWETH _weth,
+        address payable _feeCollector
+    ) Ownable(_owner) TokenCollector(_uniswapPermit2, _allowanceTarget) {
+        weth = _weth;
+        feeCollector = _feeCollector;
+    }
+
+    receive() external payable {}
+
+    function setFeeCollector(address payable _newFeeCollector) external onlyOwner {
+        if (_newFeeCollector == address(0)) revert ZeroAddress();
+        feeCollector = _newFeeCollector;
+
+        emit SetFeeCollector(_newFeeCollector);
+    }
+
+    function approveTokens(address[] calldata tokens, address[] calldata spenders) external onlyOwner {
```

Comment:
This would require owner to approve all tokens that we allow users to trade?

Reply:
Yes. Do you think that users should be able to approve tokens by themselves?

Reply:
Also, I think the `spenders` can be remove can approve 3 router for each token.

Reply:
How about adding a `approveAndSwap` function? So if a user wants to trade a new token pair, they can do it within one tx.

Reply:
Or adding a `bool` flag

Reply:
My concern is actually about owner, i.e., us, having to approve again every time users request a new token for them to trade.

And my thought was actually to automatically approve the token amount to be swapped in V2/V3 cases:
```solidity
            if (routerType == RouterType.UniversalRouter) {
                ...
            } else {
                // v2 v3 use transferFrom
                _collect(inputToken, msg.sender, address(this), inputAmount, userPermit);
                inputToken.safeApprove(routerAddr, inputAmount);
            }
```
Does this approach work or perhaps I'm missing something?

Reply:
`approveAndSwap` is probably better than a `bool` flag as user or his wallet does not have to simulate or do a pre-check before every swap. Especially when the `allowance not enough` error is not very informative for user to debug in my opinion.

Reply:
Added the `approveAndSwap` in a5e1624.